[FLINK-13361][documention] Add documentation for JDBC connector for Table API & SQL (#9802)
apache/flink: docs/dev/table/connect.md

<span class="label label-primary">Temporal Join: Sync Mode</span>

The JDBC connector allows for reading from an JDBC client.
The JDBC connector allows for writing into an JDBC client.

The connector can operate in [upsert mode](#update-modes) for exchanging UPSERT/DELETE messages with the external system using a [key defined by the query](./streaming/dynamic_tables.html#table-to-stream-conversion).

For append-only queries, the connector can also operate in [append mode](#update-modes) for exchanging only INSERT messages with the external system.

To use this connector, add the following dependency to your project:

{% highlight xml %}
<dependency>
  <groupId>org.apache.flink</groupId>
  <artifactId>flink-connector-jdbc{{ site.scala_version_suffix }}</artifactId>
  <version>{{site.version }}</version>
</dependency>
{% endhighlight %}

And must also specify JDBC library, for example, if want to use Mysql library, the following dependency to your project:

{% highlight xml %}
<dependency>
    <groupId>mysql</groupId>
    <artifactId>mysql-connector-java</artifactId>
    <version>8.0.17</version>
</dependency>
{% endhighlight %}

> **Library support:** Now, we only support mysql, derby, postgres.

Author: The "Dependencies" section show flink libraries. But mysql, derby, postgres need use version that user want.
I think we can let user decide use which version.

> **Library support:** Now, we only support mysql, derby, postgres.

Reviewer: I agree, so we can omit the version column. How about 

```
Name | Group Id | Artifact Id | JAR |
MySQL Driver | mysql | mysql-connector-java | [Download](http://central.maven.org/maven2/mysql/mysql-connector-java/) |
PostgreSQL Driver | org.postgresql | postgresql | [Download](https://jdbc.postgresql.org/download.html) |
Derby Driver | ...
```

> **Library support:** Now, we only support mysql, derby, postgres.

Author: LGTM

The connector can be defined as follows:

<div data-lang="DDL" markdown="1">
{% highlight sql %}
CREATE TABLE MyUserTable (
  ...
) WITH (
  'connector.type' = 'jdbc', -- required: specify this table type is jdbc
  
  'connector.url' = 'jdbc:derby:memory:upsert', -- required: JDBC DB url
  
  'connector.table' = 'jdbc_table_name',  -- required: jdbc table name
  
  'connector.driver' = 'driver', -- optional: jdbc driver, if not set, it will automatically guess based on the URL.

  'connector.username' = 'name', -- optional: jdbc user name and password
  'connector.password' = 'password',
  
  -- scan options, optional, used when reading from table

>   'connector.read.partition.column' = 'column_name', -- optional, name of the column used for partitioning the input.

Author: I'll make a little change and update it.

  'connector.read.partition.num' = '50', -- optional, the largest value of the last partition.
  'connector.read.partition.lower-bound' = '500', -- optional, the smallest value of the first partition.

>   'connector.read.fetch-size' = '100', -- optional, the maximum number of partitions that can be used for parallelism in table reading.

Author: This is for JDBC `Statement.setFetchSize`, I think we can just use its comments, the important thing is that it is just a hint.

> Gives the reader a hint as to the number of rows that should be fetched from the database when reading. If the value specified is zero, then the hint is ignored. The default value is zero.

>   'connector.read.fetch-size' = '100', -- optional, the maximum number of partitions that can be used for parallelism in table reading.

Reviewer: How about 

> Gives the reader a hint as to the number of rows that should be fetched from the database when reading per round trip. If the value specified is zero, then the hint is ignored. The default value is zero.

Adds "per round trip" after reading, to make it clear that no rows will be filtered if the returned size is larger than fetch size.

>   'connector.read.fetch-size' = '100', -- optional, the maximum number of partitions that can be used for parallelism in table reading.

Author: OK

  -- lookup options, optional, used in temporary join
  'connector.lookup.cache.max-rows' = '-1', -- optional, lookup cache max size, over this value, the old data will be eliminated.
  'connector.lookup.cache.ttl' = '-1', -- optional, lookup cache expire mills, over this time, the old data will expire.
  'connector.lookup.max-retries' = '3', -- optional, max retry times for jdbc connector.

  -- sink options, optional, used when writing into table
  'connector.write.flush.max-rows' = '5000', -- optional, flush max size (includes all append, upsert and delete records), 

>                                                 -- over this number of records, will flush data.

Reviewer: ```suggestion
                                                -- over this number of records, will flush data. The default value is "5000".
```

>   'connector.write.flush.interval' = '0', -- optional, flush interval mills, over this time, asynchronous threads will flush data.

Reviewer: ```suggestion
  'connector.write.flush.interval' = '2s', -- optional, flush interval mills, over this time, asynchronous threads will flush data. The default value is "0s", which means no asynchronous flush thread will be scheduled. 
```

  'connector.write.max-retries' = '3' -- optional, max retry times for jdbc connector.
)
{% endhighlight %}
</div>
</div>

**Upsert sink:** If a JDBC table is used as upsert sink, please confirm that the primary key of flink sql is the unique key of underlying database.

> **Temporary Join:** When enable cache (cacheMaxSize and cacheExpireMs not equal -1), lookup's performance will be greatly enhanced, but it may not read the latest data.

Reviewer: ```suggestion
**Temporary Join:**  JDBC connector can be used in temporal join as a lookup source. Currently, only sync lookup mode is supported. The lookup cache options (`connector.lookup.cache.max-rows` and `connector.lookup.cache.ttl`) must all be specified if any of them is specified. The lookup cache is used to improve performance of temporal join JDBC connector by querying the cache first instead of send all requests to remote database. But the returned value might not be the latest if it is from the cache. So it's a balance between throughput and correctness. 
```

> 

Reviewer: Could we add a note for writing? As default, the `connector.write.flush.interval` is `0s` and `connector.write.flush.max-rows` is `5000`, which means for low traffic queries, the buffered output rows may not be flushed to database for a long time. So the interval configuration is recommended to set.

  url: "jdbc:mysql://localhost:3306/flink-test"     # required: JDBC DB url
  table: "jdbc_table_name"        # required: jdbc table name
  driver: "com.mysql.jdbc.Driver" # optional: the class name of the JDBC driver to use to connect to this URL.
                                  # If not set, it will automatically be derived from the URL.

  username: "name"                # optional: jdbc user name and password
  password: "password"
  
  read: # scan options, optional, used when reading from table
    partition: # These options must all be specified if any of them is specified. In addition, partition.num must be specified. They
               # describe how to partition the table when reading in parallel from multiple tasks. partition.column must be a numeric,
               # date, or timestamp column from the table in question. Notice that lowerBound and upperBound are just used to decide
               # the partition stride, not for filtering the rows in table. So all rows in the table will be partitioned and returned.
               # This option applies only to reading.
      column: "column_name" # optional, name of the column used for partitioning the input.

>       num: 50               # optional, the largest value of the last partition.

Reviewer: the number of partitions?

  
  'connector.table' = 'jdbc_table_name',  -- required: jdbc table name
  
  'connector.driver' = 'com.mysql.jdbc.Driver', -- optional: the class name of the JDBC driver to use to connect to this URL. 
                                                -- If not set, it will automatically be derived from the URL.

  'connector.username' = 'name', -- optional: jdbc user name and password
  'connector.password' = 'password',
  
  -- scan options, optional, used when reading from table

  -- These options must all be specified if any of them is specified. In addition, partition.num must be specified. They
  -- describe how to partition the table when reading in parallel from multiple tasks. partition.column must be a numeric,
  -- date, or timestamp column from the table in question. Notice that lowerBound and upperBound are just used to decide
  -- the partition stride, not for filtering the rows in table. So all rows in the table will be partitioned and returned.
  -- This option applies only to reading.
  'connector.read.partition.column' = 'column_name', -- optional, name of the column used for partitioning the input.

>   'connector.read.partition.num' = '50', -- optional, the largest value of the last partition.

Reviewer: the number of partitions?

{% top %}

> <<<<<<< HEAD

Reviewer: remove
